### DISCUSSION TITLE
[lit-labs/context] Providing an object

### DISCUSSION THREAD

**Discussion by joezappie** [lit/lit#3685] [Q&A] [answered]
I'm trying out context but I'm curious how it can handle objects as the "value". For example, I want to have a timer that stores both the elapsed time and a status for if its stopped, paused, running. If I create a Timer object that is provided as the initial data, is there anyway for that object to tell the context that its value has updated? I see that the provider has a setValue command but that means remaking the object every time the timer counts up? 

Here's a lit playground example of what I'm trying to do. I have a Timer class that handles keeping track of state and intervals. A button element that shows the control buttons, a display element that shows the timer count and an app container.

[Lit Playground Example](https://lit.dev/playground/#sample=examples%2Fdirective-template-content&project=W3sibmFtZSI6Im15LWVsZW1lbnQudHMiLCJjb250ZW50IjoiXG5pbXBvcnQgeyBMaXRFbGVtZW50LCBodG1sIH0gZnJvbSAnbGl0JztcbmltcG9ydCB7IGNvbnN1bWUsIHByb3ZpZGUsIGNyZWF0ZUNvbnRleHQgfSBmcm9tICdAbGl0LWxhYnMvY29udGV4dCc7XG5pbXBvcnQgeyBjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSB9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcbmltcG9ydCB7Y2hvb3NlfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9jaG9vc2UuanMnO1xuXG5leHBvcnQgY29uc3QgdGltZXJDb250ZXh0ID0gY3JlYXRlQ29udGV4dCgndGltZXInKTtcblxuY2xhc3MgVGltZXIge1xuICBzdGF0aWMgU1RBVEVTID0ge0lETEU6MCwgUEFVU0VEOjEsIFJVTk5JTkc6Mn07XG4gIFxuICBpbnRlcnZhbCA9IG51bGw7XG4gIHN0YXRlID0gVGltZXIuU1RBVEVTLklETEU7XG4gIHZhbHVlID0gMDtcbiAgXG4gIHN0YXJ0KCkge1xuICAgIHRoaXMuc3RhdGUgPSBUaW1lci5TVEFURVMuSURMRTtcbiAgICB0aGlzLnZhbHVlID0gMDtcbiAgICB0aGlzLl9jcmVhdGVJbnRlcnZhbCgpO1xuICB9XG4gIFxuICBzdG9wKCkge1xuICAgIHRoaXMuc3RhdGUgPSBUaW1lci5TVEFURVMuSURMRTsgXG4gICAgY2xlYXJJbnRlcnZhbCh0aGlzLmludGVydmFsKTtcbiAgfVxuICBcbiAgcGF1c2UoKSB7XG4gICAgdGhpcy5zdGF0ZSA9IFRpbWVyLlNUQVRFUy5QQVVTRUQ7IFxuICAgIGNsZWFySW50ZXJ2YWwodGhpcy5pbnRlcnZhbCk7XG4gIH1cbiAgXG4gIHJlc3VtZSgpIHtcbiAgICBpZih0aGlzLnN0YXRlID09PSBUaW1lci5TVEFURVMuUEFVU0VEKSB7XG4gICAgICB0aGlzLnN0YXRlID0gVGltZXIuU1RBVEVTLlJVTk5JTkc7IFxuICAgICAgdGhpcy5fY3JlYXRlSW50ZXJ2YWwoKTtcbiAgICB9XG4gIH1cbiAgXG4gIF9jcmVhdGVJbnRlcnZhbCgpIHtcbiAgICBjbGVhckludGVydmFsKHRoaXMuaW50ZXJ2YWwpO1xuICAgIHRoaXMuaW50ZXJ2YWwgPSBzZXRJbnRlcnZhbCgoKSA9PiB7XG4gICAgICB0aGlzLnZhbHVlICs9IDE7XG4gICAgICBjb25zb2xlLmxvZyh0aGlzLnZhbHVlKTtcbiAgICB9LCAxMDAwKTtcbiAgfVxufVxuXG5AY3VzdG9tRWxlbWVudCgndGltZXItYnV0dG9ucycpXG5jbGFzcyBUaW1lckJ1dHRvbnMgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgQGNvbnN1bWUoe2NvbnRleHQ6IHRpbWVyQ29udGV4dCwgc3Vic2NyaWJlOiB0cnVlfSlcbiAgQHByb3BlcnR5KHthdHRyaWJ1dGU6IGZhbHNlfSlcbiAgdGltZXI7XG4gIFxuICByZW5kZXIoKSB7XG4gICByZXR1cm4gaHRtbGBcbiAgICAgICAke2Nob29zZSh0aGlzLnRpbWVyLnN0YXRlLCBbXG4gICAgICAgICBbVGltZXIuU1RBVEVTLklETEUsICgpID0-IGh0bWxgXG4gICAgICAgICAgIDxidXR0b24gQGNsaWNrPSR7dGhpcy50aW1lci5zdGFydCgpfT5TdGFydDwvYnV0dG9uPlxuICAgICAgICAgYF0sXG4gICAgICAgICBbVGltZXIuU1RBVEVTLlJVTk5JTkcsICgpID0-IGh0bWxgXG4gICAgICAgICAgIDxidXR0b24gQGNsaWNrPSR7dGhpcy50aW1lci5zdG9wKCl9PlN0b3A8L2J1dHRvbj5cbiAgICAgICAgICAgPGJ1dHRvbiBAY2xpY2s9JHt0aGlzLnRpbWVyLnBhdXNlKCl9PlBhdXNlPC9idXR0b24-XG4gICAgICAgICBgXSxcbiAgICAgICAgIFtUaW1lci5TVEFURVMuUEFVU0VELCAoKSA9PiBodG1sYFxuICAgICAgICAgICA8YnV0dG9uIEBjbGljaz0ke3RoaXMudGltZXIuc3RvcCgpfT5TdG9wPC9idXR0b24-XG4gICAgICAgICAgIDxidXR0b24gQGNsaWNrPSR7dGhpcy50aW1lci5yZXN1bWUoKX0-UmVzdW1lPC9idXR0b24-XG4gICAgICAgICBgXSxcbiAgICAgICBdKX1cbiAgIGA7XG4gIH1cbn1cblxuQGN1c3RvbUVsZW1lbnQoJ3RpbWVyLWRpc3BsYXknKVxuY2xhc3MgVGltZXJEaXNwbGF5IGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIEBjb25zdW1lKHtjb250ZXh0OiB0aW1lckNvbnRleHQsIHN1YnNjcmliZTogdHJ1ZX0pXG4gIEBwcm9wZXJ0eSh7YXR0cmlidXRlOiBmYWxzZX0pXG4gIHRpbWVyO1xuICBcbiAgcmVuZGVyKCkge1xuICAgcmV0dXJuIGh0bWxgPGRpdj4ke3RoaXMudGltZXIudmFsdWV9YDtcbiAgfVxufVxuXG5AY3VzdG9tRWxlbWVudCgnYXBwLWNvbnRhaW5lcicpXG5jbGFzcyBBcHBDb250YWluZXIgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgQHByb3ZpZGUoe2NvbnRleHQ6IHRpbWVyQ29udGV4dH0pXG4gIEBwcm9wZXJ0eSh7YXR0cmlidXRlOiBmYWxzZX0pXG4gIHB1YmxpYyB0aW1lciA9IG5ldyBUaW1lcigpO1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGBcbiAgICAgICAgPGhlYWRlcj5UaW1lciBBcHA8L2hlYWRlcj5cbiAgICAgICAgPG5hdmlnYXRpb24-PC9uYXZpZ2F0aW9uPlxuICAgICAgICA8Y29udGVudD5cbiAgICAgICAgICA8dGltZXItYnV0dG9ucz48L3RpbWVyLWJ1dHRvbnM-XG4gICAgICAgICAgPHRpbWVyLWRpc3BsYXk-PC90aW1lci1kaXNwbGF5PlxuICAgICAgICA8L2NvbnRlbnQ-XG4gICAgICAgIDxmb290ZXI-PC9mb290ZXI-XG4gICAgYDtcbiAgfVxufSJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPHN0eWxlPlxuICBib2R5IHsgZm9udC1mYW1pbHk6IHNhbnMtc2VyaWY7IH1cbjwvc3R5bGU-XG5cbjxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vbXktZWxlbWVudC5qc1wiPjwvc2NyaXB0PlxuXG48YXBwLWNvbnRhaW5lcj48L2FwcC1jb250YWluZXI-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d)

Also, `app-container` in this case is supposed to be generic. While its a provider for the timer context, I don't think it should have any part in the timer logic. Its just the provider as its the root container for the app. Ideally I'd also provide user contexts, organization contexts, and other global app data from `app-container`. I might be misunderstanding the use case of context however.

## TLDR
I think provider logic should be able to be separated from the element that's providing them (maybe it already is?). In this case, I'd want to provide a bunch of contexts from the root application element for data that's specifically global, such as the current user, what organization/workspace their in, settings such as user preferences or defaults. What I wouldn't want to happen is to have to make a separate element for each of these contexts and nest my entire app like:

```
<app-container>
  <timer-context>
    <user-context>
     ... Actual app code
    </user-context>
  </timer-context>
</app-container>
```

**Comment by vdegenne** [comment]
This is interesting, I had the same interrogation last time, but of course the context only reacts when your object reference changes (just like when you use a decorator like `@state`). But here are some suggestions if you need some ideas on how to implement this pattern yourself:

1) Turn your `Timer` class into a controller ([Lit/Controllers](https://lit.dev/docs/composition/controllers/)) and create accessors for both `value` and `state` so when one of the two properties changes you can call something like `this.host.updateContext()` and in your host you will have this function:

```javascript
class AppContainer {
  private _timerController = new Timer(this);

  @provide({context: timerContext})
  @property({attribute: false})
  public timerContext = { timer: this._timerController }

  updateContext() {
    // By changing the reference  your context will trigger children updates.
    // Children can then access `this.timerContext.timer`
    this.timerContext = { timer: this._timerController }
  }
}
```

Not sure if that's the best way to do that but one I can think of.

2) You could also implement your Timer as a Redux component and use `connect` from the `pwa-helpers` API....

**Reply by vdegenne** [reply]
Here's a [Lit Playground Example](https://tinyurl.com/5n6f673z)  to show case the first solution.

**Reply by vdegenne** [reply]
I am actually working on a project that can ease out this process when using Lit. I'll tell you more when I finish 😛

**Reply by joezappie** [reply]
Nice! I didn't think to use a controller for my timer class. 

The only thing I'm not a huge fan of is the fact that AppContainer still needs a function manually added to it to update the context. This becomes more complicated the more contexts you add to it, such as in my case a User context or organizational context. This however could be solved by making a simple decorator to auto add that function to the app container:

I'd love to see something built into context directly for handling this. What I'd really like to see is that context providers can be attached dynamically to an element just like a controller can. That way the Timer class itself would be a provider and I can easily remove or add it to my app container after the fact.

The docs do say that `ContextProvider` is just a reactive controller so I'm looking at just extending that class so that it can call `setValue` directly but I'd lose out on being able to use the `@provide` decorator. Maybe in the future, Lit could allow the decorator to specify your own custom provider:

```
@provide({ context: timerContext, provider: TimerProvider)
```

Interested in seeing what you come up with!

**Reply by joezappie** [reply]
Going off of that, I now have a [Playground Example](https://tinyurl.com/context-provider-extend) for extending `ContextProvider` directly. I do lose out on built in `@provide` support but easily could make my own decorator, or maybe Lit/Context would accept a PR to allow custom context providers to be used. I think that opens up a lot of doors.

**Reply by vdegenne** [reply, marked as answer]
@jrj2211 Hey,
Remember I said I would come back to you when I finish developing a solution I had in mind.
Well time has flew and I was quite busy but I remember this conversation so I am sharing this with you.
Here's the updated <a href="https://lit.dev/playground/#sample=examples%2Fdirective-template-content&gist=ee5d840ce6143d1e1398a8159cee84b2" target="_blank">Playground</a>.
I think you might be interested about it, I've managed to reduce the code to less than 90 lines, but it's not just about writing less, it's also about having a really easy way to manage states across your elements.

It's still a controller behind the scene, but a **multi-host** and **reactive** controller.
In the playground, you can see that all definition is made in the parent class `app-container` (binding the controller to all consumers in the first update method).
It works because the context is created constructor-wise, but the motivation behind `@snar/lit` was to have an easy solution to create global shared states across singleton app elements (Here's another playground to demonstrate that, it uses the `@withController` decorator : <a href="https://lit.dev/playground/#sample=examples%2Fdirective-template-content&gist=f924b5f26bc94c7d105bc28b7421ecd1" target="_blank">demo</a> which in the end is even less code.)

I've dropped the *experimental* stage, and I started to share it with people, and any feedback is appreciated.

Happy Coding!

**Reply by joezappie** [reply]
Hey @vdegenne,

Awesome work, looks great from a first glance! Definately improves the readability of all this. 

I'll start playin around with this soon, but definately will switch from my reactive controller setup to this. I'll keep you posted if I have any feedback on it.

Thanks!

**Reply by joezappie** [reply]
I'm playing around with this and the `@withController` decorator, but for some reason my elements that use the controller aren't getting rerendered when the value is updated. My ReactiveControllers `update` method is however getting called. Trying to figure out what I'm missing compared to your example.

**Reply by vdegenne** [reply]
@jrj2211 Do you have a repo so I can see what you are doing wrong (or if there is an issue with the lib)?

**Reply by joezappie** [reply]
Unfortunately I don't have a repo ATM. The only steps tho should be to create a reactive controller, create a singleton instance of said controller, then pass that into each component with @withController? Anytime any `@state` or `@property` is modified then, the lit lifecycle functions should be called in any component with the reactiveController?

I'm going to try the method of manually binding to see if its just an issue with `withController` for me.

**Reply by vdegenne** [reply]
Yes that's pretty much it, only there is no `@property` in `ReactiveObject` because that makes sense only for custom elements.
If you think you've encountered an issue please open an issue there : https://github.com/snarjs/snar/issues
I'll be more than glad to work on it 👍
